components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1358 @@
+    fn new(cfg: Config, env: Arc<Environment>, security_mgr: Arc<SecurityManager>) -> Self {
+        let context = ConnectContext {
+            cfg,
+            connector: PdConnector::new(env, security_mgr),
+        };
+        let (tx, rx) = broadcast::channel(1);
+        let core = CachedRawClientCore {
+            context,
+            latest: Arc::new(Mutex::new(None)),
+            version: Arc::new(AtomicU64::new(0)),
+            on_reconnect_tx: tx,
+        };
+        Self {
+            core: Arc::new(core),
+            on_reconnect_rx: rx,
+            cache: None,
+            cache_version: 0,
+        }
+    }
+
+    #[inline]
+    fn refresh_cache(&mut self) -> bool {
+        if self.cache_version < self.core.version.load(Ordering::Acquire) {
+            let latest = self.core.latest.lock().unwrap();
+            self.cache = (*latest).clone();
+            self.cache_version = self.core.version.load(Ordering::Relaxed);
+            true
+        } else {
+            false
+        }
+    }
+
+    #[inline]
+    async fn publish_cache(&mut self) {
+        let latest_version = {
+            let mut latest = self.core.latest.lock().unwrap();
+            *latest = self.cache.clone();
+            let _ = self.core.on_reconnect_tx.send(());
+            self.core.version.fetch_add(1, Ordering::Relaxed) + 1
+        };
+        debug_assert!(self.cache_version < latest_version);
+        self.cache_version = latest_version;
+    }
+
+    #[inline]
+    async fn wait_for_a_new_client(
+        rx: &mut broadcast::Receiver<()>,
+        current_version: u64,
+        latest_version: &AtomicU64,
+    ) -> bool {
+        let deadline = std::time::Instant::now() + REQUEST_TIMEOUT;
+        loop {
+            if GLOBAL_TIMER_HANDLE
+                .timeout(Compat::new(Box::pin(rx.recv())), deadline)
+                .compat()
+                .await
+                .is_ok()
+            {
+                if current_version < latest_version.load(Ordering::Acquire) {
+                    return true;
+                }
+            } else {
+                return false;
+            }
+        }
+    }
+
+    /// Refreshes the local cache with latest client, then waits for the
+    /// connection to be ready.
+    /// The connection must be available if this function returns `Ok(())`.
+    async fn wait_for_ready(&mut self) -> Result<()> {
+        self.refresh_cache();
+        if self.cache.is_none() {
+            if !Self::wait_for_a_new_client(
+                &mut self.on_reconnect_rx,
+                self.cache_version,
+                self.core.version.as_ref(),
+            )
+            .await
+            {
+                return Err(box_err!("Connection is not initialized in time"));
+            }
+            assert!(self.refresh_cache());
+            assert!(self.cache.is_some());
+        }
+        select! {
+            r = self.core.grab_a_channel().wait_for_connected(REQUEST_TIMEOUT).fuse() => {
```

Comment:
As in most case, it should be connected, so you can create a short path by calling `check_connectivity_state`.